sql: disallow a column being backfilled in an expression

vivekmenezes opened the pull request:
Before this change a column being backfilled could be used
in any arbitrary expression and the system would panic when
it used such a column in an index constraint expression.

fixes #18316

Release note (bug fix): fix crash when using a column being
backfilled in an index constraint

vivekmenezes committed:
sql: disallow a column being backfilled in an expression

Before this change a column being backfilled could be used
in any arbitrary expression and the system would panic when
it used such a column in an index constraint expression.

fixes #18316

Release note (bug fix): fix crash when using a column being
backfilled in an index constraint

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21308)
<!-- Reviewable:end -->

vivekmenezes commented:
I hate passing around these booleans to make decisions and I'm happy to consider something else. It took me quite some time to make this small change.

knz commented:
I understand the need to do this in `findColHelper` and the passing around of booleans.
See my suggestions below. Foremost I am concerned about extending `ResultColumns`.

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sem/tree/var_name.go, line 155 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21308#-L2Jz8nLYc6fjaLbqFqH:-L2Jz8nMZc5IzUBQv9rZ:b-466xre) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fb8321d9496ed86cdfbfe7225a4f4e3a825c715/pkg/sql/sem/tree/var_name.go#L155)):*
> ```Go
> 	// TODO(vivek): Do not artificially create such internal expressions
> 	// when scanning columns for an UPDATE/DELETE.
> 	InternalUseOnly bool
> ```

The name `InternalUseOnly` does not clearly convey what this boolean is for.
Given how you are using it, perhaps `BackfillTarget`?

---

*[pkg/sql/sqlbase/result_columns.go, line 37 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21308#-L2JyFi4RdtksPHl24Wb:-L2JyFi539s8OFJXk_AE:b-bsy6hx) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fb8321d9496ed86cdfbfe7225a4f4e3a825c715/pkg/sql/sqlbase/result_columns.go#L37)):*
> ```Go
> 	// and is needed to construct internal data like a Column-Family or
> 	// an Index.
> 	IsBeingBackfilled bool
> ```

I understand the need to mark columns being backfilled in a data source.
However extending the ResultColumn type is very invasive: it creates memory overhead on the common path, to serve the needs of a (relatively) uncommon operation.

As an alternate solution I would recommend adding a bitmap of columns being backfilled (or, generally, undergoing mutation) in the `sql.dataSourceInfo` struct. Populate this bitmap upon instantiating a data source for a table, and then access it in `findColHelper()`.

---

*[pkg/sql/sqlbase/structured.go, line 2145 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21308#-L2JzoJ8bMejPf1LQgP3:-L2JzoJ8bMejPf1LQgP4:b-ipw76p) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fb8321d9496ed86cdfbfe7225a4f4e3a825c715/pkg/sql/sqlbase/structured.go#L2145)):*
> ```Go
> 
> // ColumnsSelectors generates Select expressions for cols.
> func ColumnsSelectors(cols []ColumnDescriptor) tree.SelectExprs {
> ```

`UPDATE` may not be the only user of this method in the future. I would recommend adding a boolean argument to this function (e.g. `backfillTarget bool`), and use it to assign the new field below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21308#-:-L2JzfD-xRP9sgOXxceh:b-cuc888)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 2 of 10 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/sem/tree/var_name.go, line 155 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21308#-L2Jz8nLYc6fjaLbqFqH:-L2MFJ8BUvu4LvawrQzb:bn631my) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fb8321d9496ed86cdfbfe7225a4f4e3a825c715/pkg/sql/sem/tree/var_name.go#L155)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The name `InternalUseOnly` does not clearly convey what this boolean is for.
Given how you are using it, perhaps `BackfillTarget`?
</blockquote></details>

renamed to the more explicit ForUpdateOrDelete

---

*[pkg/sql/sqlbase/result_columns.go, line 37 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21308#-L2JyFi4RdtksPHl24Wb:-L2MPGdPnZ8xTRjChCC_:b-mq1a6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fb8321d9496ed86cdfbfe7225a4f4e3a825c715/pkg/sql/sqlbase/result_columns.go#L37)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I understand the need to mark columns being backfilled in a data source.
However extending the ResultColumn type is very invasive: it creates memory overhead on the common path, to serve the needs of a (relatively) uncommon operation.

As an alternate solution I would recommend adding a bitmap of columns being backfilled (or, generally, undergoing mutation) in the `sql.dataSourceInfo` struct. Populate this bitmap upon instantiating a data source for a table, and then access it in `findColHelper()`.
</blockquote></details>

I'd like to do that but don't know how to go about it in a sensible way. Ideally we can express a threshold H above which all result columns are backfill columns. Unfortunately the signature of newSourceInfoForSingleTable will have to be extended and then it will need to be properly called from a number of callers.

The memory overhead of this boolean is not too bad.

---

*[pkg/sql/sqlbase/structured.go, line 2145 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21308#-L2JzoJ8bMejPf1LQgP3:-L2MFOR3dxRJcbqC2Xob:bm75lag) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fb8321d9496ed86cdfbfe7225a4f4e3a825c715/pkg/sql/sqlbase/structured.go#L2145)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`UPDATE` may not be the only user of this method in the future. I would recommend adding a boolean argument to this function (e.g. `backfillTarget bool`), and use it to assign the new field below.
</blockquote></details>

changed to pass in boolean forUpdateOrDelete

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21308)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 2 of 10 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/result_columns.go, line 37 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21308#-L2JyFi4RdtksPHl24Wb:-L2MSflu5AloIoD33l19:b-4dwz76) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fb8321d9496ed86cdfbfe7225a4f4e3a825c715/pkg/sql/sqlbase/result_columns.go#L37)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

I'd like to do that but don't know how to go about it in a sensible way. Ideally we can express a threshold H above which all result columns are backfill columns. Unfortunately the signature of newSourceInfoForSingleTable will have to be extended and then it will need to be properly called from a number of callers.

The memory overhead of this boolean is not too bad.
</blockquote></details>

I would recommend having a new field in `dataSourceInfo` which, when unset/nil, means "no backfill column".
Then you can create a new method separate from `newSourceInfoForSingleTable`, which takes that new parameter to indicate the backfill columns.
Then use that method when instantiating a datasourceinfo for a table scan (one of the leaf cases in `getSource`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21308)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 2 of 10 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/result_columns.go, line 37 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21308#-L2JyFi4RdtksPHl24Wb:-L2MSvMhTueOBtoZSfJZ:bphdjgc) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fb8321d9496ed86cdfbfe7225a4f4e3a825c715/pkg/sql/sqlbase/result_columns.go#L37)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I would recommend having a new field in `dataSourceInfo` which, when unset/nil, means "no backfill column".
Then you can create a new method separate from `newSourceInfoForSingleTable`, which takes that new parameter to indicate the backfill columns.
Then use that method when instantiating a datasourceinfo for a table scan (one of the leaf cases in `getSource`).
</blockquote></details>

Also there is a method somewhere to join two `dataSourceInfo` together (for joins). This will also need to concatenate the two backfill bitmaps.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21308)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 1 of 9 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/sqlbase/result_columns.go, line 37 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/21308#-L2JyFi4RdtksPHl24Wb:-L2MhCugIUApeRrWH-aa:b-pjpw82) ([raw file](https://github.com/cockroachdb/cockroach/blob/8fb8321d9496ed86cdfbfe7225a4f4e3a825c715/pkg/sql/sqlbase/result_columns.go#L37)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Also there is a method somewhere to join two `dataSourceInfo` together (for joins). This will also need to concatenate the two backfill bitmaps.
</blockquote></details>

I've changed this to keep a count of the number of columns that are backfill columns and use that to determine which column is a backfill column.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21308)*
<!-- Sent from Reviewable.io -->

knz commented:
Yes this looks good. Thank you!

@jordanlewis mind having a 2nd look? That I'm not missing anything.

---

Reviewed 9 of 9 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21308#-:-L2Mp2pFnBdoZiWLHlrg:b5orckk)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Any concerns with me merging this?

knz commented:
None from me.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 2 of 9 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/21308#-:-L2_W7i8qvxOJjGHd5oo:bnfp4nl)*
<!-- Sent from Reviewable.io -->
